File: solutions/typescript/arrays-and-hashing/two-sum.ts
/**
 * 1. Two Sum — https://leetcode.com/problems/two-sum/
 *
 * Given `nums` and `target`, return the indices of the two numbers adding up to `target`.
 *
 * Both approaches below are registered and reachable from the API via `?approach=`, so you
 * can benchmark them against each other over the same input without changing any code.
 */

import { NoSolutionError } from "../errors";
import { defineProblem } from "../define-problem";

export const SLUG = "two-sum";

/**
 * Check every pair. Time O(n^2), space O(1).
 *
 * Worth keeping: it is the baseline that makes the hash-map version's space cost look like a
 * bargain, and on tiny inputs it is genuinely faster because it never allocates.
 */
export function twoSumBruteForce(nums: readonly number[], target: number): number[] {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i]! + nums[j]! === target) return [i, j];
    }
  }
  throw new NoSolutionError(SLUG, `No two entries of nums sum to ${target}.`);
}

/**
 * One pass, remembering every value already seen. Time O(n), space O(n).
 *
 * `Map`, not a plain object: object keys are coerced to strings, so `-0` and `0` would
 * collide and numeric keys would be re-parsed on every lookup. `Map` keeps real number keys
 * with SameValueZero semantics. Python's `dict` and PHP's array both sidestep this — it is a
 * genuinely JavaScript-only trap.
 */
export function twoSumHashMap(nums: readonly number[], target: number): number[] {
  const seen = new Map<number, number>();
  for (let index = 0; index < nums.length; index++) {
    const value = nums[index]!;
    const complement = seen.get(target - value);
    if (complement !== undefined) return [complement, index];
    seen.set(value, index);
  }
  throw new NoSolutionError(SLUG, `No two entries of nums sum to ${target}.`);
}

export const twoSum = defineProblem(SLUG, {
  "brute-force": twoSumBruteForce,
  "hash-map": twoSumHashMap,
});

File: solutions/python/arrays_and_hashing/two_sum.py
"""1. Two Sum — https://leetcode.com/problems/two-sum/

Given `nums` and `target`, return the indices of the two numbers adding up to `target`.

Both approaches below are registered and reachable from the API via `?approach=`, so you can
benchmark them against each other over the same input without changing any code.
"""

from __future__ import annotations

from neetcode_core.errors import NoSolutionError
from neetcode_core.registry import solution

SLUG = "two-sum"


@solution(SLUG, approach="brute-force")
def two_sum_brute_force(nums: list[int], target: int) -> list[int]:
    """Check every pair.

    Time O(n^2), space O(1). Worth keeping: it is the baseline that makes the hash-map
    version's space cost look like a bargain, and on tiny inputs it is genuinely faster
    because it never allocates.
    """
    for i in range(len(nums)):
        for j in range(i + 1, len(nums)):
            if nums[i] + nums[j] == target:
                return [i, j]
    raise NoSolutionError(SLUG, f"No two entries of nums sum to {target}.")


@solution(SLUG, approach="hash-map")
def two_sum_hash_map(nums: list[int], target: int) -> list[int]:
    """One pass, remembering every value already seen.

    Time O(n), space O(n). The trick is to look *backwards* for the complement instead of
    forwards for the partner: by the time the second number of the pair is the current
    element, the first one is guaranteed to already be in `seen`.
    """
    seen: dict[int, int] = {}
    for index, value in enumerate(nums):
        complement = target - value
        if complement in seen:
            return [seen[complement], index]
        seen[value] = index
    raise NoSolutionError(SLUG, f"No two entries of nums sum to {target}.")
